library/nist_kw.c

```diff
@@ -0,0 +1,650 @@
+        return( ret );
+
+    if( ( ret = mbedtls_cipher_setkey( &ctx->cipher_ctx, key, keybits,
+                               isWrap ? MBEDTLS_ENCRYPT : MBEDTLS_DECRYPT ) ) != 0 )
+    {
+        return( ret );
+    }
+
+    return( 0 );
+}
+
+/*
+ * Free context
+ */
+void mbedtls_nist_kw_free( mbedtls_nist_kw_context *ctx )
+{
+    mbedtls_cipher_free( &ctx->cipher_ctx );
+    mbedtls_platform_zeroize( ctx, sizeof( mbedtls_nist_kw_context ) );
+}
+
+/*
+ * KW / KWP - AE
+ */
+int mbedtls_nist_kw_wrap( mbedtls_nist_kw_context *ctx, int mode,
+                     const unsigned char *input, size_t in_len,
+                     unsigned char *output, size_t* out_len )
+{
+    int ret = 0;
+    size_t semiBlocks = 0;
+    size_t s ;
+    size_t olen, i, padlen = 0;
+    uint64_t t = 0;
+    unsigned char outBuff[ KW_SEMIBLOCK_LENGTH * 2 ];
+    unsigned char inBuff[ KW_SEMIBLOCK_LENGTH * 2 ];
+    unsigned char* R2 = output + KW_SEMIBLOCK_LENGTH;
+    unsigned char* A = output ;
+    const uint32_t  ICV1[] = NIST_KW_ICV1;
+    const uint32_t  ICV2[] = NIST_KW_ICV2;
+
+/*
+ * generate the String to work on
+ */
+    if( mode == MBEDTLS_KW_MODE_KW )
+    {
+/*
+ * According to SP800-38F Table 1, plaintext length size limits
+ * for KW is 2 to 2^54-1 semiblocks
+ */
+        if( ( in_len < 8 )
+#if SIZE_MAX > UINT_MAX
+                ||( in_len > 0x1FFFFFFFFFFFFF8 )
+#endif
+           )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+
+        memcpy( output, ICV1, KW_SEMIBLOCK_LENGTH );
+        memcpy( output + KW_SEMIBLOCK_LENGTH, input, in_len );
+    }
+    else
+    {
+        /*
+         * According to SP800-38F Table 1, plaintext length size limits
+         * for KWP is 1 to 2^32-1 octets
+         */
+        if( ( in_len < 1 )
+#if SIZE_MAX > UINT_MAX
+           || (in_len > 0xFFFFFFFF )
+#endif
+          )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+        if( (in_len % 8) != 0 )
+        {
+            padlen = ( 8 - ( in_len % 8 ) );
+        }
+
+        memcpy( output, &ICV2, KW_SEMIBLOCK_LENGTH / 2 );
+        PUT_UINT32_BE( ( in_len & 0xffffffff ), output, KW_SEMIBLOCK_LENGTH / 2 );
+
+        memcpy( output + ( KW_SEMIBLOCK_LENGTH / 2 ) + sizeof(uint32_t), input, in_len );
+        memset( output + ( KW_SEMIBLOCK_LENGTH / 2 ) + sizeof(uint32_t) + in_len, 0, padlen );
+    }
+    semiBlocks = ( ( in_len + padlen ) / KW_SEMIBLOCK_LENGTH ) + 1;
+
+    s = 6 * ( semiBlocks -1 );
+
+    if( ( mode == MBEDTLS_KW_MODE_KWP )
+            && ( in_len <= KW_SEMIBLOCK_LENGTH ) )
+    {
+        memcpy( inBuff, output, 16 );
+        ret = mbedtls_cipher_update( &ctx->cipher_ctx, inBuff, 16, output, &olen);
+        if( ret != 0 )
+            goto cleanup;
+
+        *out_len = olen;
+    }
+    else
+    /*
+     * Do the wrapping function W
+     */
+    {
+        if( semiBlocks < 3 )
```

Comment:
What is `3`? Consider making a local constant or enum.

Reply:
It's a value from the standard. It's the minimal number of semiblocks allowed for wrapping